csharp/src/Drivers/Apache/Hive2/HiveServer2Connection.cs

```diff
@@ -664,50 +628,62 @@ internal static async Task PollForResponseAsync(TOperationHandle operationHandle
         {
-            TGetInfoReq req = new()
+            return TraceActivity(activity =>
             {
-                SessionHandle = SessionHandle ?? throw new InvalidOperationException("session not created"),
-                InfoType = infoType,
-            };
+                TGetInfoReq req = new()
+                {
+                    SessionHandle = SessionHandle ?? throw new InvalidOperationException("session not created"),
+                    InfoType = infoType,
+                };
 
-            CancellationToken cancellationToken = ApacheUtility.GetCancellationToken(QueryTimeoutSeconds, ApacheUtility.TimeUnit.Seconds);
-            try
-            {
-                TGetInfoResp getInfoResp = Client.GetInfo(req, cancellationToken).Result;
-                if (getInfoResp.Status.StatusCode == TStatusCode.ERROR_STATUS)
+                CancellationToken cancellationToken = ApacheUtility.GetCancellationToken(QueryTimeoutSeconds, ApacheUtility.TimeUnit.Seconds);
+                try
                 {
-                    throw new HiveServer2Exception(getInfoResp.Status.ErrorMessage)
-                        .SetNativeError(getInfoResp.Status.ErrorCode)
-                        .SetSqlState(getInfoResp.Status.SqlState);
-                }
+                    TGetInfoResp getInfoResp = Client.GetInfo(req, cancellationToken).Result;
+                    ApacheUtility.HandleThriftResponse(getInfoResp.Status, GetResponseHandlers(activity));
 
-                return getInfoResp.InfoValue.StringValue;
-            }
-            catch (Exception ex) when (ExceptionHelper.IsOperationCanceledOrCancellationRequested(ex, cancellationToken))
-            {
-                throw new TimeoutException("The metadata query execution timed out. Consider increasing the query timeout value.", ex);
-            }
-            catch (Exception ex) when (ex is not HiveServer2Exception)
-            {
-                throw new HiveServer2Exception($"An unexpected error occurred while running metadata query. '{ex.Message}'", ex);
-            }
+                    return getInfoResp.InfoValue.StringValue;
+                }
+                catch (Exception ex) when (ExceptionHelper.IsOperationCanceledOrCancellationRequested(ex, cancellationToken))
+                {
+                    throw new TimeoutException("The metadata query execution timed out. Consider increasing the query timeout value.", ex);
+                }
+                catch (Exception ex) when (ex is not HiveServer2Exception)
+                {
+                    throw new HiveServer2Exception($"An unexpected error occurred while running metadata query. '{ex.Message}'", ex);
+                }
+            });
         }
 
-        public override void Dispose()
+        protected override void Dispose(bool disposing)
         {
-            if (_client != null && SessionHandle != null)
+            TraceActivity(activity =>
```

Comment:
`activity` isn't used until HandleThriftResponse. Consider factoring the network operation into a separate method and putting the `TraceActivity` into that method and keeping the `Dispose` small and more readable.